src/allocation_engine/README.md

```diff
@@ -0,0 +1,170 @@
+
+ ```rust
+/// A closed interval range [min, max] used to describe a
+/// memory slot that will be assigned to a device by the VMM.
+/// This structure represents the key of the Node object in
+/// the interval tree implementation.
+pub struct Range {
+    pub min: u64,
+    pub max: u64,
+}
+
+/// Node state for interval tree nodes.
+///
+/// Valid state transition:
+/// - None -> Free: IntervalTree::insert(key, NodeState::Free)
+/// - Free -> Allocated: IntervalTree::insert(key, NodeState::Allocated)
+/// - Allocated -> Free: IntervalTree::free()
+/// - * -> None: IntervalTree::delete()
+pub enum NodeState {
+    /// Node is free.
+    Free,
+    /// Node is allocated.
+    Allocated,
+}
+
+/// Internal tree node to implement interval tree.
+pub(crate) struct InnerNode {
+    /// Interval handled by this node.
+    pub(crate) key: Range,
+    /// State of the node, this can be `Free` or `Allocated`
+    pub(crate) node_state: NodeState,
+    /// Optional left child of current node.
+    pub(crate) left: Option<Box<InnerNode>>,
+    /// Optional right child of current node.
+    pub(crate) right: Option<Box<InnerNode>>,
+    /// Cached height of the node.
+    pub(crate) height: u64,
+    /// Cached maximum valued covered by this node.
+    pub(crate) max_key: u64,
+}
+```
+
+## Usage
+
+The concept of Interval Tree may seem complicated, but using vm-allocator to do
+resource allocation and release is simple and straightforward.
+
+To use the `IntervalTree` implementation as an address allocator one should first
+create an interval tree object and give an address space as a root node.
+
+```rust
+pub struct AddressAllocator {
+    base: u64,
+    end: u64,
+    interval_tree: IntervalTree,
+}
+
+impl AddressAllocator {
+    pub fn new(base: u64, size: u64) -> std::result::Result<Self, Error> {
+        let interval_tree = IntervalTree::new();
+        let end = base.checked_add(size - 1)?;
+        let mut address_allocator = AddressAllocator {
+            base: base,
+            end: end,
+            interval_tree: interval_tree,
+        };
+
+        address_allocator
+            .interval_tree
+            .insert(Range::new(base, end)?, NodeState::Free)?;
+        Ok(address_allocator)
+    }
+```
+
+After, the user should create a constraint with the size for the resource, optionally
+ the constraint could also contain the maximum, minimum and alignment for the
+ constraint.
+
+```rust
+ pub fn allocate(
+        &mut self,
+        _address: Option<u64>,
+        size: u64,
+        align_size: Option<u64>,
+        alloc_policy: AllocPolicy,
+    ) -> Result<Range> {
+        let constraint = Constraint::new(size).align(alignment).policy(alloc_policy);
+        // some other code here
+        if logical_condition {
+            let key = self.interval_tree.find_candidate(&constaint);
+            self.interval_tree.insert(key, NodeState::Allocated);
+        }
+    }
+```
+
+![Allocation example](/images/allocation_example.png)
```

Comment:
I don't really understand the idea of the image. What do the big boxes represent? What about the arrows? Maybe we can sync offline to clarify that a bit.